src/program.cpp

```diff
@@ -353,13 +353,17 @@ std::vector<argument> program::eval(parameter_map params) const
 
     if(trace_level > 0)
     {
+        std::unordered_map<instruction_ref, std::string> ins_names;
+        // get instruction names
+        this->print(ins_names, [](auto, auto) {});
```

Comment:
We should fill the map with the instruction print outs:

```cpp
std::unordered_map<instruction_ref, std::string> ins_out;
this->print([&](auto x, auto ins_names) {
    std::stringstream ss;
    instruction::print(ss, x, ins_names);
    ins_out[x] = ss.str();
});
```

Then when we want to print out we can do `std::cout << "Run instruction: " << ins_out.at(ins) << std::endl;`.

Reply:
We also should add a `print` overload that doesnt take a map of names:

```cpp
void program::print(
    const std::function<void(instruction_ref, std::unordered_map<instruction_ref, std::string>)>&
        print_func) const
{
    std::unordered_map<instruction_ref, std::string> names;
    this->print(names, print_func);
}
```

Reply:
fixed